add cluster-launch-installer-src template

ecordell opened the pull request:
Differences from `cluster-launch-installer-e2e`:

- Uses `LOCAL_IMAGE_SRC` instead of `IMAGE_TESTS`
- Removes gingko commands

ecordell commented:
/assign @bbguimaraes

bbguimaraes commented on ci-operator/templates/openshift/installer/cluster-launch-installer-src.yaml:
```diff
@@ -0,0 +1,357 @@
+        mkdir -p "${HOME}"
+
+        # wait for the router namespace
+        FOUND_KUBECONFIG=
```
This is missing these changes: https://github.com/openshift/release/pull/2062

ecordell commented:
@bbguimaraes fixed - must've based this off of an older version of `-e2e`, sorry about that.

bbguimaraes commented on ci-operator/templates/openshift/installer/cluster-launch-installer-src.yaml:
```diff
@@ -0,0 +1,355 @@
+- name: IMAGE_INSTALLER
+  required: true
+- name: LOCAL_IMAGE_SRC 
+  required: true
```
```diff
-- name: IMAGE_TESTS
+- name: LOCAL_IMAGE_SRC
+  required: true
+- name: IMAGE_CLI
```

`IMAGE_CLI` needs to be a parameter of the template or we won't get its value from ci-operator.

bbguimaraes commented on ci-operator/templates/openshift/installer/cluster-launch-installer-src.yaml:
```diff
@@ -0,0 +1,355 @@
+      - |
+        #!/bin/bash
+        set -euo pipefail
+        export PATH=/usr/libexec/origin:$PATH
```
```diff
-        export PATH=/usr/libexec/origin:$PATH
+        export PATH=/tmp/shared:$PATH
```

The binary is placed in a different location in this template.

bbguimaraes commented on ci-operator/templates/openshift/installer/cluster-launch-installer-src.yaml:
```diff
@@ -0,0 +1,355 @@
+        export PATH=/usr/libexec/origin:$PATH
+        trap 'touch /tmp/shared/exit' EXIT
+        trap 'kill $(jobs -p); exit 0' TERM
+        cp "$(which oc)" /tmp/shared/
```
```diff
-        cp "$(which oc)" /tmp/shared/
```

That's already done by the init container.

bbguimaraes commented on ci-operator/templates/openshift/installer/cluster-launch-installer-src.yaml:
```diff
@@ -0,0 +1,355 @@
+          cp /tmp/cluster/ssh-privatekey ~/.ssh/kube_openstack_rsa || true
+        fi
+        mkdir -p /tmp/output
+        cd /tmp/output
```
```diff
-        mkdir -p /tmp/output
-        cd /tmp/output
```

We can't `cd` out of the repository's directory or `${TEST_COMMAND}` will likely fail.

bbguimaraes commented on ci-operator/templates/openshift/installer/cluster-launch-installer-src.yaml:
```diff
@@ -0,0 +1,355 @@
+      - -c
+      - |
+        #!/bin/sh
+        trap 'rc=$?; if test "${rc}" -ne 0; then touch /tmp/exit; fi; exit "${rc}"' EXIT
```
```diff
-        trap 'rc=$?; if test "${rc}" -eq 0; then touch /tmp/setup-success; else touch /tmp/exit; fi; exit "${rc}"' EXIT
+        trap 'rc=$?; if test "${rc}" -ne 0; then touch /tmp/exit; fi; exit "${rc}"' EXIT
```

This one's still outstanding from #2062.

ecordell replied:
To clarify - I *should* be checking `/tmp/setup-success`, right?

ecordell commented:
@bbguimaraes Addressed comments

Full diff of `e2e` to `src`:

```diff
15c15,17
< - name: IMAGE_TESTS
---
> - name: LOCAL_IMAGE_SRC 
>   required: true
> - name: IMAGE_CLI
60c62,71
< 
---
>     initContainers:
>     - name: cli
>       image: ${IMAGE_CLI}
>       volumeMounts:
>       - name: shared-tmp
>         mountPath: /tmp/shared
>       command:
>       - cp
>       - /usr/bin/oc
>       - /tmp/shared/oc
65c76
<       image: ${IMAGE_TESTS}
---
>       image: ${LOCAL_IMAGE_SRC}
92c103
<         export PATH=/usr/libexec/origin:$PATH
---
>         export PATH=/tmp/shared:$PATH
96,98c107
< 
<         cp "$(which oc)" /tmp/shared/
< 
---
>         
180,183c189
< 
<         mkdir -p /tmp/output
<         cd /tmp/output
< 
---
>         
193,211d198
<         # TODO: the test binary should really be a more structured command - most of these flags should be
<         #       autodetected from the running cluster.
<         # TODO: bump nodes up to 40 again
<         function run-tests() {
<           if [[ -n "${TEST_FOCUS:-}" ]]; then
<             ginkgo -v -noColor -nodes="${TEST_PARALLELISM:-30}" $( which extended.test ) -- \
<               -ginkgo.focus="${TEST_FOCUS}" -ginkgo.skip="${TEST_SKIP:-"\\[local\\]"}" \
<               -e2e-output-dir /tmp/artifacts -report-dir /tmp/artifacts/junit \
<               -test.timeout=2h ${PROVIDER_ARGS-} || rc=$?
<           fi
<           if [[ -n "${TEST_FOCUS_SERIAL:-}" ]]; then
<             ginkgo -v -noColor -nodes=1 $( which extended.test ) -- \
<               -ginkgo.focus="${TEST_FOCUS_SERIAL}" -ginkgo.skip="${TEST_SKIP_SERIAL:-"\\[local\\]"}" \
<               -e2e-output-dir /tmp/artifacts -report-dir /tmp/artifacts/junit/serial \
<               -test.timeout=2h ${PROVIDER_ARGS-} || rc=$?
<           fi
<           exit ${rc:-0}
<         }
< 
268c255
<         trap 'rc=$?; if test "${rc}" -eq 0; then touch /tmp/setup-success; else touch /tmp/exit; fi; exit "${rc}"' EXIT
---
>         trap 'rc=$?; if test "${rc}" -ne 0; then touch /tmp/exit; fi; exit "${rc}"' EXIT

```

wking commented:
With the small difference between `e2e` and `src`, is this something we could be controlling with an environment variable or some such to keep all the code in `cluster-launch-installer-e2e.yaml`?

ecordell commented:
@wking I don't use `Template`s enough to speak authoritatively, but I don't think there are good ways to conditionally swap which parameter gets used in one place. The main concern is swapping `image` from `${IMAGE_TESTS}` to `${LOCAL_IMAGE_SRC}`.

For this PR I'm just copying the pattern for pre-4.0 templates.

bbguimaraes commented:
@wking: as @ecordell said, the fundamental difference between the two templates is which image to use for the `test` container, which is not something that can be switched with an environment variable.

@ecordell: comparing that diff and the diff between the openshift-ansible templates, the only thing missing is:

https://github.com/openshift/release/blob/e284409070dc9197faa743dab107530f53fb0c44/ci-operator/templates/cluster-launch-src.yaml#L123

Add that and we should be done with this.

ecordell commented:
@bbguimaraes Added the `ARTIFACT_DIR` env var

bbguimaraes commented:
/lgtm
